### PR TITLE
Correlated bai

### PR DESCRIPTION
**Pull request by choltz95**
To get more familiar with the api for best arm identification in fixed conf setting I implement the alg & environment from https://arxiv.org/abs/1710.05552 for [single player] best arm selection in linear/correlated bandit environment. Still need to debug some of the functions, but it looks like it's working okay at the moment. A little slower compared to lilUCB. Could use your comments.

Screen Shot 2020-02-23 at 4 10 44 PM
Screen Shot 2020-02-23 at 4 10 32 PM

### CODE REVIEW

**Comment by choltz95** [comment]
For the correlated algorithms I think I can move the em arm initialization to utils `_model_init`:
```
    # initalize em arms repr
    for k in range(self._arm_num):
      self._em_arms[k].action = self._bandit.actions[k]
```

**Comment by Alanthink** [comment]
> For the correlated algorithms I think I can move the em arm initialization to utils `_model_init`:
> 
> ```
>     # initalize em arms repr
>     for k in range(self._arm_num):
>       self._em_arms[k].action = self._bandit.actions[k]
> ```

Yes, I think this is a good idea. Can you fix the warnings generated by pep8? Also, can you add the ref in the readme file?

File: learners/bestarmid/fixconf/correlatedlearner/lingape.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
+import math
```
Comment (Alanthink) [inline learners/bestarmid/fixconf/correlatedlearner/lingape.py]: unused import

File: learners/bestarmid/fixconf/correlatedlearner/lingape.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+import numpy as np
+
+from absl import flags
+
+from bandits.arms import EmArm
```
Comment (Alanthink) [inline learners/bestarmid/fixconf/correlatedlearner/lingape.py]: unused import

File: learners/bestarmid/fixconf/correlatedlearner/lingape.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  def __select_direction(self, t):
+    i_t = np.argmax([np.dot(a.action, self.__th) for a in self._em_arms])
+    gap_conf = [self.__Delta(j, i_t) +
+                self.__beta(j, i_t)
+                for j in range(self._arm_num)]
+    j_t = np.argmax(gap_conf)
+    B_t = max(gap_conf)
+    return i_t, j_t, B_t
+
+  def __mat_norm(self, x, A):
```
Comment (Alanthink) [inline learners/bestarmid/fixconf/correlatedlearner/lingape.py]: this function can be put into `utils.py`

File: learners/bestarmid/fixconf/correlatedlearner/lingape.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+  def __beta(self, i, j):
+    x_diff = self._em_arms[i].action - self._em_arms[j].action
+    return self.__mat_norm(x_diff, np.linalg.inv(self.__A)) * self.__C_n
+
+  def __Delta(self, i, j):
+    x_diff = self._em_arms[i].action - self._em_arms[j].action
+    return np.dot(x_diff, self.__th)
+
+  def __select_direction(self, t):
```
Comment (Alanthink) [inline learners/bestarmid/fixconf/correlatedlearner/lingape.py]: `t` argument is not used

File: learners/bestarmid/fixconf/correlatedlearner/utils.py [hunk 5]
```diff
@@ -14,6 +16,10 @@ class CorrelatedLearner(FixConfBAILearner):
   def name(self):
     pass
 
+
+  def _mat_norm(self, x, A):
```
Comment (Alanthink) [inline learners/bestarmid/fixconf/correlatedlearner/utils.py]: Can you put this outside class and add `__all__ = ['FixConfBAILearner', 'mat_norm']` before class definition?

Comment (choltz95) [reply learners/bestarmid/fixconf/correlatedlearner/utils.py]: Does this look correct [in correlatedlearner/utils.py]?

Screen Shot 2020-02-24 at 12 02 38 PM

Comment (Alanthink) [reply learners/bestarmid/fixconf/correlatedlearner/utils.py]: Yes, you are right. Two lines before and after `mat_norm` definition.

**Comment by choltz95** [comment]
Note: only implemented greedy selection strategy

**Pull request merged**